lib/NormalModule.js

```diff
@@ -69,260 +40,385 @@ function contextify(options, request) {
 				return callback();
 			}
+
+			try {
+				this.parser.parse(this._source.source(), {
+					current: this,
+					module: this,
+					compilation: compilation,
+					options: options
+				});
+			} catch(e) {
+				const source = this._source.source();
+				this.error = new ModuleParseError(this, source, e);
+				this.markModuleAsErrored();
+				return callback();
+			}
+			return callback();
+		});
+	}
+
+	source(dependencyTemplates, outputOptions, requestShortener) {
+		let hash = require("crypto").createHash("md5");
+		this.updateHash(hash);
+		hash = hash.digest("hex");
+		if(this._cachedSource && this._cachedSource.hash === hash) {
+			return this._cachedSource.source;
+		}
+		const _source = this._source;
+		if(!_source) return new RawSource("throw new Error('No source available');");
+		const source = new ReplaceSource(_source);
+		this._cachedSource = {
+			source: source,
+			hash: hash
+		};
+		const topLevelBlock = this;
+
+		function doDep(dep) {
+			const template = dependencyTemplates.get(dep.constructor);
+			if(!template) throw new Error("No template for dependency: " + dep.constructor.name);
+			template.apply(dep, source, outputOptions, requestShortener, dependencyTemplates);
 		}
-		try {
-			_this.parser.parse(_this._source.source(), {
-				current: _this,
-				module: _this,
-				compilation: compilation,
-				options: options
+
+		function doVariable(availableVars, vars, variable) {
+			const name = variable.name;
+			const expr = variable.expressionSource(dependencyTemplates, outputOptions, requestShortener);
+
+			function isEqual(v) {
+				return v.name === name && v.expression.source() === expr.source();
+			}
+			if(availableVars.some(isEqual)) return;
+			vars.push({
+				name: name,
+				expression: expr
 			});
-		} catch(e) {
-			var source = _this._source.source();
-			return setError(_this.error = new ModuleParseError(_this, source, e));
 		}
-		return callback();
-	});
-
-	function setError(err) {
-		_this.meta = null;
-		if(_this.error) {
-			_this.errors.push(_this.error);
-			_this._source = new RawSource("throw new Error(" + JSON.stringify(_this.error.message) + ");");
-		} else {
-			_this._source = new RawSource("throw new Error('Module build failed');");
+
+		function doBlock(availableVars, block) {
+			block.dependencies.forEach(doDep);
+			const vars = [];
+			if(block.variables.length > 0) {
+				block.variables.forEach(doVariable.bind(null, availableVars, vars));
+				const varNames = [];
+				const varExpressions = [];
+				let varStartCode = "";
+				let varEndCode = "";
+
+				const emitFunction = function emitFunction() {
+					if(varNames.length === 0) return;
+					varStartCode += "/* WEBPACK VAR INJECTION */(function(" + varNames.join(", ") + ") {";
+					// exports === this in the topLevelBlock, but exports do compress better...
+					varEndCode = (topLevelBlock === block ? "}.call(" + (topLevelBlock.exportsArgument || "exports") + ", " : "}.call(this, ") +
```

Comment:
Could you make this line more readable?  i. e. by moving some parts into local variables.